taskcluster/firefox_tv_taskgraph/job.py

```diff
@@ -0,0 +1,175 @@
+from taskgraph.util import path
+from taskgraph.util.schema import Schema, taskref_or_string
+from voluptuous import Required, Optional
+from six import text_type
+
+from pipes import quote as shell_quote
+
+secret_schema = {
+    Required("name"): text_type,
+    Required("path"): text_type,
+    Required("key"): text_type,
+    Optional("json"): bool,
+}
+
+dummy_secret_schema = {
+    Required("content"): text_type,
+    Required("path"): text_type,
+    Optional("json"): bool,
+}
+
+gradlew_schema = Schema({
+    Required("using"): "gradlew",
+    Optional("pre-gradlew"): [[text_type]],
+    Required("gradlew"): [text_type],
+    Optional("post-gradlew"): [[text_type]],
+    # Base work directory used to set up the task.
+    Required("workdir"): text_type,
+    Optional("use-caches"): bool,
+    Optional("secrets"): [secret_schema],
+    Optional("dummy-secrets"): [dummy_secret_schema],
+})
+
+run_commands_schema = Schema({
+    Required("using"): "run-commands",
+    Optional("pre-commands"): [[text_type]],
+    Required("commands"): [[taskref_or_string]],
+    Required("workdir"): text_type,
+    Optional("use-caches"): bool,
+    Optional("secrets"): [secret_schema],
+    Optional("dummy-secrets"): [dummy_secret_schema],
+})
+
+
+@run_job_using("docker-worker", "run-commands", schema=run_commands_schema)
+def configure_run_commands_schema(config, job, taskdesc):
+    run = job["run"]
+    pre_commands = run.pop("pre-commands", [])
+    pre_commands += [
+        _generate_dummy_secret_command(secret) for secret in run.pop("dummy-secrets", [])
+    ]
+    pre_commands += [
+        _generate_secret_command(secret) for secret in run.get("secrets", [])
+    ]
+
+    all_commands = pre_commands + run.pop("commands", [])
+
+    run["command"] = _convert_commands_to_string(all_commands)
+    _inject_secrets_scopes(run, taskdesc)
+    _set_run_task_attributes(job)
+    configure_taskdesc_for_run(config, job, taskdesc, job["worker"]["implementation"])
+
+
+@run_job_using("docker-worker", "gradlew", schema=gradlew_schema)
+def configure_gradlew(config, job, taskdesc):
+    run = job["run"]
+    worker = taskdesc["worker"] = job["worker"]
+
+    worker.setdefault("env", {}).update({
+        "ANDROID_SDK_ROOT": path.join(
+            run["workdir"], worker["env"]["MOZ_FETCHES_DIR"], "android-sdk-linux"
+        )
+    })
+
+    run["command"] = _extract_gradlew_command(run)
+    _inject_secrets_scopes(run, taskdesc)
+    _set_run_task_attributes(job)
+    configure_taskdesc_for_run(config, job, taskdesc, job["worker"]["implementation"])
+
+
+def _extract_gradlew_command(run):
+    pre_gradle_commands = run.pop("pre-gradlew", [])
+    pre_gradle_commands += [
+        _generate_dummy_secret_command(secret) for secret in run.pop("dummy-secrets", [])
+    ]
+    pre_gradle_commands += [
+        _generate_secret_command(secret) for secret in run.get("secrets", [])
+    ]
+
+    gradle_command = ["./gradlew"] + run.pop("gradlew")
```

Comment:
Looking at this, I'm curious if there is any value in making this gradle-specific, and having the pre- and post- bits. (particularly as I see one of the post commands is also using gradle).

Reply:
Hmmm, that's a great question! Let's see.

The `gradlew` kind was first introduced in [August 2019](https://github.com/mozilla-mobile/reference-browser/pull/737/files#diff-81790a570b45107d44ea76233a7754e8R33). `post-gradlew` was added [the following month](https://github.com/mozilla-mobile/reference-browser/pull/892/files#diff-24c9b9fc54c62f21ea7365d5b268f925R54). `pre-gradlew` was then added [3 weeks later](https://github.com/mozilla-mobile/fenix/pull/5488/files#diff-d4885d206919bcc0e772ccdb9618c071R18).

`run-commands` first appeared in [November 2019](https://github.com/mozilla-mobile/fenix/pull/6517/files#diff-d4885d206919bcc0e772ccdb9618c071R42), in order to support non-gradle tasks.

We could convert `gradlew` flavors into `run-commands` since the former grew organically. I actually wondered if I could combine [that singular post-command](https://github.com/mozilla-mobile/firefox-tv/pull/2984/files#diff-ec565f108a3a09f10c613b7dea5ee557R38) into the big gradlew list. Although, I didn't want to break anything so I preferred to keep 2 separate commands. 

One thing I like about keeping `gradlew` is: https://github.com/mozilla-mobile/reference-browser/pull/1117/files#diff-81790a570b45107d44ea76233a7754e8R99-R105. By keeping this flavor around, we allow each `gradlew` call to be rewired to use specific maven repositories. This is something we cannot easily do with `run-commands`.

To sum up, I'd be in favor of keeping `gradlew` and try to merge the one post-command that still uses `gradlew` in a potential followup. For the context, firefox-tv is a deprecated project. I've just been backporting important taskgraph patches. I currently don't see this gradlew merge as important. How does that sound to you @tomprince?